go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,456 @@
+				// This is used for nodeport
+				vips = make([]string, 0, len(vips))
+				for _, vip := range config.vips {
+					if vip == "node" {
+						vips = append(vips, node.nodeIPs...)
+					} else {
+						vips = append(vips, vip)
+					}
+				}
+
+				for _, vip := range vips {
+					// build switch rules
+					targets := switchV4Targets
+					if strings.Contains(vip, ":") {
+						targets = switchV6Targets
+					}
+
+					switchRules = append(switchRules, ovnlb.LBRule{
+						Source:  ovnlb.Addr{IP: vip, Port: config.inport},
+						Targets: targets,
+					})
+
+					// For shared gateway, there is also a per-router rule
+					// with targets that *may* be different
+					if globalconfig.Gateway.Mode == "shared" {
+						targets := routerV4targets
+						if strings.Contains(vip, ":") {
+							targets = routerV6targets
+						}
+						rule := ovnlb.LBRule{
+							Source:  ovnlb.Addr{IP: vip, Port: config.inport},
+							Targets: targets,
+						}
+
+						// in other words, is this ExternalTrafficPolicy=local?
+						if config.routerLocalMode {
+							noSNATRouterRules = append(noSNATRouterRules, rule)
+						} else {
+							routerRules = append(routerRules, rule)
+						}
+					}
+				}
+			}
+
+			// If switch and router rules are identical, coalesce
+			if reflect.DeepEqual(switchRules, routerRules) && len(switchRules) > 0 && node.gatewayRouterName != "" {
+				out = append(out, ovnlb.LB{
+					Name:        makeLBName(service, proto, "node_router+switch_"+node.name),
+					Protocol:    string(proto),
+					ExternalIDs: eids,
+					Opts:        lbOpts(service),
+					Routers:     []string{node.gatewayRouterName},
+					Switches:    []string{node.switchName},
+					Rules:       routerRules,
+				})
+			} else {
+				if len(routerRules) > 0 && node.gatewayRouterName != "" {
+					out = append(out, ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_router_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Routers:     []string{node.gatewayRouterName},
+						Rules:       routerRules,
+					})
+				}
+				if len(noSNATRouterRules) > 0 && node.gatewayRouterName != "" {
+					lb := ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_local_router_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Routers:     []string{node.gatewayRouterName},
+						Rules:       noSNATRouterRules,
+					}
+					lb.Opts.SkipSNAT = true
+					out = append(out, lb)
+				}
+
+				if len(switchRules) > 0 {
+					out = append(out, ovnlb.LB{
+						Name:        makeLBName(service, proto, "node_switch_"+node.name),
+						Protocol:    string(proto),
+						ExternalIDs: eids,
+						Opts:        lbOpts(service),
+						Switches:    []string{node.switchName},
+						Rules:       switchRules,
+					})
+				}
+			}
+		}
+	}
+
+	merged := mergeLBs(out)
```

Comment:
Right, so we only create per-node load-balancers in the following cases:
- NodePort
- host-network endpoints
- ExternalTrafficPolicy=local

For NodePort services, every node will have a different vip, so there will be no merging going on. For host-network endpoints and ExternalTrafficPolicy=local, the load balancer targets depend on whether or not there are endpoints on that node. 

Rather than try and detect every possible case, it's easier to write the logic in a simple 1:1 manner, then merge equivalent configurations. That helps ensure correctness. And, since we sort everything before merging, we're guaranteed a stable merge between iterations.